test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/AbstractService.java

```diff
@@ -120,38 +127,82 @@ public String getID() {
 
-        final Map<String, Object> pravegaSpec = ImmutableMap.<String, Object>builder().put("controllerReplicas", controllerCount)
-                                                                                      .put("segmentStoreReplicas", segmentStoreCount)
-                                                                                      .put("debugLogging", true)
-                                                                                      .put("cacheVolumeClaimTemplate", pravegaPersistentVolumeSpec)
-                                                                                      //.put("options", props.getProperties())
-                                                                                      .put("options", props)
-                                                                                      .put("image",
-                                                                                       getImageSpec(DOCKER_REGISTRY + PREFIX + "/" + PRAVEGA_IMAGE_NAME, PRAVEGA_VERSION))
-                                                                                      .put("tier2", tier2Spec())
-                                                                                      .build();
-        return ImmutableMap.<String, Object>builder()
-                .put("apiVersion", "pravega.pravega.io/v1alpha1")
-                .put("kind", CUSTOM_RESOURCE_KIND_PRAVEGA)
-                .put("metadata", ImmutableMap.of("name", PRAVEGA_ID, "namespace", NAMESPACE))
-                .put("spec", ImmutableMap.builder().put("zookeeperUri", zkLocation)
-                                         .put("bookkeeper", bookkeeperSpec)
-                                         .put("pravega", pravegaSpec)
-                                         .build())
-                .build();
+        String bookkeeperImg = DOCKER_REGISTRY + PREFIX + "/" + BOOKKEEPER_IMAGE_NAME;
+        String pravegaImg = DOCKER_REGISTRY + PREFIX + "/" + PRAVEGA_IMAGE_NAME;
+
+        if (isLaterOperator) {
+
+            final Map<String, Object> bookkeeperSpec = ImmutableMap.<String, Object>builder().put("image", ImmutableMap.of("repository", bookkeeperImg))
+                    .put("replicas", bookieCount)
+                    .put("resources", getResources("2000m", "5Gi", "1000m", "3Gi"))
+                    .put("storage", ImmutableMap.builder()
+                            .put("indexVolumeClaimTemplate", bkPersistentVolumeSpec)
+                            .put("ledgerVolumeClaimTemplate", bkPersistentVolumeSpec)
+                            .put("journalVolumeClaimTemplate", bkPersistentVolumeSpec)
+                            .build())
+                    .put("autoRecovery", true)
+                    .build();
+
+            final Map<String, Object> pravegaSpec = ImmutableMap.<String, Object>builder().put("controllerReplicas", controllerCount)
```

Comment:
To avoid code repetition and improve modularity you could have spec builder classes for different operator versions (>=0.4.0 vs older) and depending on the builder used the right specs would get generated and populated here.

Reply:
As most of the code is the same in both cases with only one or two changes required for each, I don't think that would be necessary.

Reply:
@SrishT It looks like you haven;t understood what I;m suggesting above. 
I'm suggesting builder classes based on version range, in the same lines as what Raul had suggested recently. It is definitely a better and more clean way of doing things, through not absolutely critical to this fix.